### PR TITLE
feat(router): add `start_redirection` api for three_ds flow in v2

### PR DESCRIPTION
**Pull request by sai-harsha-vardhan**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

add start_redirection api for three_ds flow in v2
1. store the redirection_data sent by the connector flows in payment_attempt
2. introduce next_action block in confirm intent response and add support for three_ds flow of card payments
3. add start_redirection api to support three ds flow for card payments in v2

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Tested Manually
1. create a payment with authentication_type: three_ds

CURL
```
curl --location '{{BASE_URL}}/v2/payments/create-intent' \
--header 'api-key: {{API_KEY}}' \
--header 'X-Profile-Id: {{PROFILE_ID}}' \
--header 'Content-Type: application/json' \
--data '{
    "amount_details": {
        "order_amount": 100,
        "currency": "USD"
    },
    "authentication_type": "three_ds"
}'
```
Response
```
{
    "id": "12345_pay_0192dd8aae0b7050af8ff8849326fefb",
    "amount_details": {
        "order_amount": 100,
        "currency": "USD",
        "shipping_cost": null,
        "order_tax_amount": null,
        "skip_external_tax_calculation": "Skip",
        "skip_surcharge_calculation": "Skip",
        "surcharge_amount": null,
        "tax_on_surcharge": null
    },
    "client_secret": "12345_pay_0192dd8aae0b7050af8ff8849326fefb_secret_0192dd8aae0b7050af8ff896254521c8",
    "merchant_reference_id": null,
    "routing_algorithm_id": null,
    "capture_method": "automatic",
    "authentication_type": "three_ds",
    "billing": null,
    "shipping": null,
    "customer_id": null,
    "customer_present": "Present",
    "description": null,
    "return_url": null,
    "setup_future_usage": "on_session",
    "apply_mit_exemption": "Skip",
    "statement_descriptor": null,
    "order_details": null,
    "allowed_payment_method_types": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "payment_link_enabled": "Skip",
    "payment_link_config": null,
    "request_incremental_authorization": "default",
    "expires_on": "2024-10-30T13:23:26.251Z",
    "frm_metadata": null,
    "request_external_three_ds_authentication": "Skip"
}
```

2. confirm a payment and get the next_action block for redirect_url

CURL
```
curl --location '{{BASE_URL}}/v2/payments/{{PAYMENT_ID}}/confirm-intent' \
--header 'Content-Type: application/json' \
--header 'X-Profile-Id: {{PROFILE_ID}}' \
--header 'X-Client-Secret: {{CLIENT_SECRET}}' \
--data '{
    "payment_method_data": {
            "card": {
                "card_number": "4242424242424242",
                "card_exp_month": "01",
                "card_exp_year": "2097",
                "card_holder_name": "John Doe",
                "card_cvc": "100"
            }
    },
    "payment_method_type": "card",
    "payment_method_subtype": "credit",
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "125.0.0.1"
    }
}'
```
Response:
```
{
    "id": "12345_pay_0192dd8aae0b7050af8ff8849326fefb",
    "status": "requires_customer_action",
    "amount": {
        "order_amount": 100,
        "currency": "USD",
        "shipping_cost": null,
        "order_tax_amount": null,
        "skip_external_tax_calculation": "Skip",
        "skip_surcharge_calculation": "Skip",
        "surcharge_amount": null,
        "tax_on_surcharge": null,
        "net_amount": 100,
        "amount_to_capture": null,
        "amount_capturable": 0,
        "amount_captured": null
    },
    "connector": "checkout",
    "client_secret": "12345_pay_0192dd8aae0b7050af8ff8849326fefb_secret_0192dd8aae0b7050af8ff896254521c8",
    "created": "2024-10-30T13:08:26.251Z",
    "payment_method_data": null,
    "payment_method_type": "card",
    "payment_method_subtype": "credit",
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/v2/payments/12345_pay_0192dd8aae0b7050af8ff8849326fefb/start_redirection?merchant_id=cloth_seller1_CLporg83u4dFG54nP7z3"
    },
    "connector_transaction_id": "pay_zwj4ebatyseeppxcqv7fsjkz7u",
    "connector_reference_id": null,
    "merchant_connector_id": "mca_uxXfZhmnm6sRI6fOgDiw",
    "browser_info": null,
    "error": null
}
```

3. click on the redirect_url to see the redirection going to connector page
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: v2_migrations/2024-08-28-081838_update_v2_primary_key_constraints/up.sql [hunk 1]
```diff
@@ -130,4 +130,9 @@ ALTER TABLE payment_intent ALTER COLUMN session_expiry SET NOT NULL;
 -- This migration is to make fields optional in payment_intent table
 ALTER TABLE payment_intent ALTER COLUMN active_attempt_id DROP NOT NULL;
 
+-- This migration is to make fields optional in payment_intent table
+ALTER TABLE payment_intent
```
Comment (Narayanbhat166) [inline v2_migrations/2024-08-28-081838_update_v2_primary_key_constraints/up.sql:134]: This is already run on the above line right?

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -4604,6 +4607,20 @@ pub struct PaymentsConfirmIntentResponse {
...
+#[cfg(feature = "v2")]
+pub struct PaymentStartRedirectionRequest {
+    /// Global Payment ID
+    pub id: id_type::GlobalPaymentId,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+#[cfg(feature = "v2")]
+pub struct PaymentStartRedirectionParams {
+    /// The identifier for the Merchant Account.
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:4620]: We should include profile-id also

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -5927,6 +5929,71 @@ pub async fn payment_external_authentication(
     ))
 }
 
+#[instrument(skip_all)]
+#[cfg(feature = "v2")]
+pub async fn payment_start_redirection(
```
[the remark this answers is not in the document]
Comment (jarnura) [reply crates/router/src/core/payments.rs:5934]: @sai-harsha-vardhan can you confirm this, are we using any connector related calls, if using those why we are using also.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments.rs:5934]: we are not doing any connector call, but some information related to payment instrument is required for a connector. So we fetch these details from the temp locker.

[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/core/payments.rs:5934]: @SamraatBansal can you please confirm this?

File: crates/router/src/core/payments.rs [hunk 4]
```diff
@@ -5927,6 +5929,71 @@ pub async fn payment_external_authentication(
...
+
+    let payment_intent = db
+        .find_payment_intent_by_id(key_manager_state, &req.id, &key_store, storage_scheme)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+
+    utils::when(
+        payment_intent.status != storage_enums::IntentStatus::RequiresCustomerAction,
+        || {
+            Err(errors::ApiErrorResponse::PaymentUnexpectedState {
```
[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/core/payments.rs:5957]: I'll add a TODO for this

File: crates/router/src/core/payments.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/core/payments.rs:5934]: I can see that we're using card_cvc in bluesnap redirect form, is it needed? @SamraatBansal 
```
RedirectForm::BlueSnap {
            payment_fields_token,
        } => {
            let card_details = if let Some(PaymentMethodData::Card(ccard)) = payment_method_data {
                format!(
                    "var saveCardDirectly={{cvv: \"{}\",amount: {},currency: \"{}\"}};",
                    ccard.card_cvc.peek(),
                    amount,
                    currency
                )
            } else {
                "".to_string()
            };
```

[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/core/payments.rs:5934]: we are currently not supporting payment method data fetch in this API, as it is an open API without authentication. We will need to verify the connector implementation when necessary, as sending card_cvc to the client is not the appropriate way to handle this flow @Narayanbhat166

File: crates/hyperswitch_domain_models/src/payments.rs [hunk 5]
```diff
@@ -98,6 +98,17 @@ impl PaymentIntent {
     pub fn get_id(&self) -> &id_type::GlobalPaymentId {
         &self.id
     }
+
+    #[cfg(feature = "v2")]
+    pub fn create_start_redirection_url(&self, base_url: &str) -> String {
```
Comment (Narayanbhat166) [inline crates/hyperswitch_domain_models/src/payments.rs:103]: this should ideally give out `url` type

File: crates/router/src/core/payments.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments.rs:5934]: Okay sure, makes sense to not expose card details in this api.

File: crates/router/src/core/payments.rs [hunk 6]
```diff
@@ -5927,6 +5929,76 @@ pub async fn payment_external_authentication(
...
+
+    let storage_scheme = merchant_account.storage_scheme;
+
+    let payment_intent = db
+        .find_payment_intent_by_id(key_manager_state, &req.id, &key_store, storage_scheme)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+
+    // validate session expiry
+    helpers::authenticate_client_secret(Some(&payment_intent.client_secret), &payment_intent)?;
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments.rs:5951]: we are passing the same client secret here? should we not take the client secret in the api query params?

File: crates/router/src/core/payments.rs [hunk 7]
```diff
@@ -5927,6 +5929,76 @@ pub async fn payment_external_authentication(
...
+    let form: RedirectForm = serde_json::from_value(redirection_data.expose()).map_err(|err| {
+        logger::error!(error = ?err, "Failed to deserialize redirection data");
+        errors::ApiErrorResponse::InternalServerError
+    })?;
+
+    Ok(services::ApplicationResponse::Form(Box::new(
+        services::RedirectionFormData {
+            redirect_form: form,
+            payment_method_data: None,
+            amount: payment_intent.amount_details.order_amount.to_string(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments.rs:5996]: where do we need amount? should this be net amount from payment attempt?

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

File: crates/router/src/services/authentication.rs [hunk 8]
```diff
@@ -1142,6 +1142,65 @@ where
...
+                self.publishable_key.as_str(),
+            )
+            .await
+            .map_err(|e| {
+                if e.current_context().is_db_not_found() {
+                    e.change_context(errors::ApiErrorResponse::Unauthorized)
+                } else {
+                    e.change_context(errors::ApiErrorResponse::InternalServerError)
+                }
+            })?;
```
Comment (ThisIsMani) [inline crates/router/src/services/authentication.rs:1177]: NIT: Can you `to_not_found_response` instead.

File: crates/router/src/services/authentication.rs [hunk 9]
```diff
@@ -1142,6 +1142,65 @@ where
...
+
+        let profile = state
+            .store()
+            .find_business_profile_by_profile_id(
```
Comment (ThisIsMani) [inline crates/router/src/services/authentication.rs:1181]:
```suggestion
          .find_business_profile_by_profile_id_merchant_id(
```

**Review by tsdk02 — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]
You can create an issue and pick these up in a different PR, approving for now.

**Review by Narayanbhat166 — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

**Review by tsdk02 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**